### DISCUSSION TITLE
Unable to use Parameter value in qmd document when passed through command line

### DISCUSSION THREAD

**Discussion by sunvitD** [quarto-dev/quarto-cli#11587] [Q&A] [answered]
### Description

````qmd
---
title: "Test File"
format: html
html:
embed-resources: true
execute:
  echo: False
jupyter: python3
---

# Title

Print this in report
```{python}
foo
```

````

Command used:
```
quarto render test.qmd -P foo:5 --output test_out.html
```

Getting **Error**:
```
Starting python3 kernel...Done

Executing 'test.quarto_ipynb'
  Cell 1/1: ''...ERROR:

An error occurred while executing the following cell:
------------------
foo
------------------


---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 foo

NameError: name 'foo' is not defined
```

**Doubt:** 

Is this not the correct way to call or evaluate parameter value in qmd python cell ?

Do I need to use it like in `R` `params$foo`. 

Would really appreciate if you could show me the correct way to use it. Thanks !!!

**Comment by mcanouil** [comment]
See the documentation: https://quarto.org/docs/computations/parameters.html

Providing a value to the parameter via CLI flag still require your document to follow the parameter syntax for the engine you are using.

**Reply by sunvitD** [reply]
I have changed the qmd file slightly as per the documentation but it is still giving me same error:
````
---
title: "Test File"
format: html
html:
embed-resources: true
execute:
  echo: False
jupyter: python3
---

# Title

Print this in report
```{python}
#| tags: [parameters]

foo
```
````

**Error:**
```
Starting python3 kernel...Done

Executing 'test-8.quarto_ipynb'
  Cell 1/2: ''...ERROR:

An error occurred while executing the following cell:
------------------
foo
------------------


---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 foo

NameError: name 'foo' is not defined
```

I am trying to call the parameter value here but in documentation examples are mostly related to assigning the value. So not sure how its need to be called in qmd file when parameter is used via CLI flag . 

Document examples below
![image](https://github.com/user-attachments/assets/98012324-8fe1-47f4-afaf-439f7804b679)

![image](https://github.com/user-attachments/assets/f1818b5c-895e-4ba0-8241-82ce12f20414)

If you don't mind then could you provide a **quarto example** just to get it right this time ?

**Reply by mcanouil** [reply, marked as answer]
You did not follow the documentation and your YAML frontmatter is wrong, *i.e.*, indentation and the weird top-level `html` key.

<table>
<tr><th>Input</th><th>Output</th></tr>
<tr><td>

````qmd
---
title: "Test File"
format: html
embed-resources: true
execute:
  echo: false
jupyter: python3
---

# Title

Print this in report

```{python}
#| tags: [parameters]
foo = "default"
```

```{python}
# Using the parameter
print(foo)
```

````

</td><td>

<img width="1624" alt="image" src="https://github.com/user-attachments/assets/e0b5358d-8164-431a-b509-5cc8b04001f6">

</td></tr>
</table>

**Reply by sunvitD** [reply]
I had no idea about using `foo = "default"`. Thanks alot @mcanouil specially for being there even on weekend. Really appreciate it !!

**Reply by mcanouil** [reply]
> I had no idea about using `foo = "default"`. Thanks alot @mcanouil specially for being there even on weekend. Really appreciate it !!

FYI, That's written in the documentation and in the screenshot you took:

> To parameterize a document, designate a cell with the tag parameters and provide appropriate default values:

You are welcome ;)

**Reply by sunvitD** [reply]
Well the document wasn't really clear enough for me in this case .. it might be for others. If you hadn't shown me the example of `foo = "default` I wouldn't have figure out. But thanks again.

Cheers !!

**Reply by sunvitD** [reply]
how can we send a list in parameters flag?

I tried:
`quarto render test.qmd -P foo:[5,8,3] --output test_out.html` this is printing as `default` literal and not the values in the qmd rendered file. 

`quarto render test.qmd -P foo:list(5,8,3) --output test_out.html` also returns the same.

**Reply by mcanouil** [reply]
You need to use YAML syntax. https://yaml.org

For complex and/or numerous parameters, there is another flag allowing you to provide a YAML file with all parameters.

**Reply by sunvitD** [reply]
Thank you for the quick reply, so if I have `data.yml` file:
```
- bar: 6.0
  foo: 1
  ham: a
- bar: 7.0
  foo: 2
  ham: b
- bar: 8.0
  foo: 3
  ham: c
```

command will become this:
`quarto render test.qmd --execute-params data.yml --output test_out.html`

Then below `qmd` file should be able to work as it is right or do i need to change anything if `paramter` flags are changing?

````
---
title: "Test File"
format: html
embed-resources: true
execute:
  echo: false
jupyter: python3
---

# Title

Print this in report

```{python}
#| tags: [parameters]
foo = "default"
```

```{python}
# Using the parameter
print(foo)
```
````

Apologies for so many questions,  I am new to `Quarto` and have tried to look for blogs and video tutorials but was not able to find examples on `parameters` and `yaml` file using `Quarto`.

**Reply by mcanouil** [reply]
Yes this should work but I would recommend you set an appropriate default value for the parameter.
"Default" was only an example.
If you want your default to be at least of the same type possibly with type checking on the parameter in your code.

**Reply by sunvitD** [reply]
I am not sure if I have understood this correctly, so the `"default"` is not here for assigning value injected from yaml file using flags but it is to define the type? Please correct me if I am wrong.

My goal is to read these 3 variables values and then convert it into a `dataframe` and give them datatypes there:
````
```{python}
#| tags: [parameters]
foo = "default"
bar="default"
ham="default
```
````
And if I have used below code then that would have also worked (if I ignore the type just to understand this):
````
```#| tags: [parameters]
foo
bar
ham
```

```{python}
print(foo)
```
```{python}
print(bar)
```
````

I think I got it now, so I need to assign data type like in below chunk:
````
```#| tags: [parameters]
foo = "int" # instead of "default"
```
````

**Reply by mcanouil** [reply]
Set real values for the default values.

If `foo` is supposed to be numeric, don't set `foo = "default"`, set `foo = 123`.
Set appropriate default values.

**Reply by sunvitD** [reply]
Thanks for the reply and the help. Understood the logic now. Really Really Appreciate your help on my naive questions !!
